pkg/factory/risingwave_object_factory.go

```diff
@@ -72,7 +91,15 @@ func (f *RisingWaveObjectFactory) namespace() string {
 }
 
 func (f *RisingWaveObjectFactory) isObjectStorageS3() bool {
-	return f.risingwave.Spec.Storages.Object.S3 != nil
+	return f.risingwave.Spec.Storages.Object.S3 != nil && len(f.risingwave.Spec.Storages.Object.S3.Endpoint) == 0
+}
+
+func (f *RisingWaveObjectFactory) isObjectStorageS3Compatible() bool {
+	return f.risingwave.Spec.Storages.Object.S3 != nil && len(f.risingwave.Spec.Storages.Object.S3.Endpoint) > 0
```

Comment:
We may run into problems here. If e.g. `f.risingwave.Spec.Storages` for some reason is `nil` we run into a nullptr dereference.

Reply:
>We may run into problems here. If e.g. f.risingwave.Spec.Storages for some reason is nil we run into a nullptr dereference.

Yes, you're right! 

> Or do we just count on the validating webhook preventing that? What if someone changed the webhook?

Yes, we do the validation in the webhook, so it's guaranteed to be non-nil. But if someone breaks the webhook, then the snippets above will panic.

> We could solve this with pointer getter functions like [in this post](https://stackoverflow.com/questions/28215031/test-for-nil-values-in-nested-stucts). Do you think it is worth the effort?

I think it's good to have something like the getter functions only if we know what to do when the value is missing. But I don't think we have a better way than to panic, and it's somewhat identical to the nil-deref panic here. Correct me if I'm wrong. 🤔

Reply:
Yes and no. Let's say we would implement It like this: 

```go
func (f *RisingWaveObjectFactory) getRisingwave() *risingwavev1alpha1.RisingWave {
	if f == nil {
		return nil
	}
	return f.risingwave
}

func (f *RisingWaveObjectFactory) isObjectStorageS3Compatible() bool {
	s3 := f.getRisingwave().getSpec().getStorages().getObject().getS3()
	return s3 != nil && len(s3.getEndpoint()) > 0
}
```

We would not run into a `nullptr` deref, but right now we use `isObjectStorageS3Compatible` only in `hummockConnectionStr`, which will then throw `panic("unrecognized storage type")`. 

I would still argue that above approach is the cleaner solution. It enables us to use `isObjectStorageS3Compatible` in a way that does not cause a panic. 

What do you think? In any ways, I believe it is a separate issue and should not be fixed in this PR

Reply:
Hmm, right. But for this one, it can not be implemented without modifying the `api` package. Because it's something like the chain access in Java, but we don't have such a pattern builtin in Go. And it seems hard to maintain it without a code generator.

```go
f.getRisingwave().getSpec().getStorages().getObject().getS3()
```

Reply:
Fair enough. Let's put in on the back burner. Maybe we will find a good solution for this eventually. I opened an issue for this so we do not forget about it https://github.com/risingwavelabs/risingwave-operator/issues/256